### DISCUSSION TITLE
Welcome to OSMD Discussions!

### DISCUSSION THREAD

**Discussion by sschmidTU** [opensheetmusicdisplay/opensheetmusicdisplay#950] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about (how to do something specific with OSMD, etc).
  * Share ideas.
  * Share your OSMD-related projects
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with OSMD,
  or your OSMD-related question.

**Comment by infojunkie** [comment]
Hello! My name is Karim and I am interested to build Web-based music applications. More specifically, I am interested in the following topics:

- Rendering and playing leadsheets including auto-generated accompaniments
- Collaborative music notebooks (Google Docs + Jupyter for music)
- Music blogging (MathJax for music)
- Music internationalization (supporting non-Western music systems in tunings / notation / fonts / sounds / rhythms / etc.- Unicode for music)
- Music theory visualization 

You can find more info about the above on my [GitHub projects](https://github.com/infojunkie?tab=projects) and on [my blog](https://blog.karimratib.me/).

I am hoping to adopt and contribute to a robust Web music engraver that will support the above use cases and I hope OSMD will be the one :heart:

**Reply by sschmidTU** [reply]
Hey Karim, nice to hear more about your aspirations with and beyond OSMD!

Did you know we support half sharp (quartertone sharp) and flat accidentals, e.g. for arabic music?
<img src="https://user-images.githubusercontent.com/33069673/98159543-12fb5900-1edd-11eb-9555-1e8a0d6a90b6.png" height="100">
<img src="https://user-images.githubusercontent.com/33069673/98159472-f828e480-1edc-11eb-993e-522495b34b4f.png" height="100">
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/903#issuecomment-721933642

Hope to hear more from you in the future!

Also, feel free to check out [our Discord server](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/927#issuecomment-766828339), you can present your projects there, which others have already done:
<img src="https://user-images.githubusercontent.com/33069673/106205269-8cde6e00-61be-11eb-9783-e5589d6de551.png" height="300">

**Reply by infojunkie** [reply]
Thanks Simon, yes I was thrilled that OSMD supports microtonal accidentals. I worked on supporting them in VexFlow a few years ago!

**Reply by sschmidTU** [reply]
Oh wow, thanks to you as well then! Right, microtonal. I'm always thinking about the quarter tone piano with microtonals ;)

**Comment by loganknecht** [comment]
# Introduction

Hello everyone!

🎹 💪 I'm `Logan Knecht` and I'm the creator of Piano Gym 💪 🎹 

Our goal at Piano Gym is to help people learn Piano and music theory with as little friction as possible using flash cards!
The main idea being that we just want you to hit the gym and do your reps everyday!

We are currently in beta and trying to fix some of the last remaining issues before we kick into gear to start reaching out to communities and make more flash cards.

Oh! And, we're free!

You can find our website here:
https://pianogym.com/

And a quick demo for it here!
https://www.youtube.com/watch?v=EMmSM4RB1NI

We're also available on social media!
- Instagram: https://www.instagram.com/Piano_Gym
- Patreon: https://patreon.com/pianogym
- Reddit: https://reddit.com/r/PianoGym
- Twitch: https://twitch.tv/piano_gym
- Twitter: https://twitter.com/PianoGym

-----

The reason `Piano Gym` uses OSMD is because we need a way to render sheet music in order to display music xml files. Furthermore we want to interact with the sheet music in order to animate different portions so that we can provide users graded feedback while they play with a Piano.

Previously we were using Vexflow to render music, but our focus was to move as close as possible to what the actual experience of playing sheet music is like. We saw OSMD and fell in love!


# Features That Have Been Invaluable From OSMD

- Loading Music XML files
- Remote loading by url
- Start measure selection with access to whole piece
- End measure selection with access to whole piece
- Library updates
- Iterator logic to navigate through the music sheet piece
- Team's accessibility and responses
 
# Features That Would Help

- A helper library to interrogate sheet music
    - Something that doesn't require the DOM to load it
    - Think of this as a preflight check to confirm data or pull out information in a configuration menu before loading OSMD to render
- Ability to animate notation
    - Think making notes disappear with a small firework
- Continuous cursor movement versus discrete movement (aka smooth scrolling)
    - I have had to solve this separately
- Ability to hook into note play timing, using a window so that you can start detecting 300ms before note and after note (would be used to grade notes)
- Helper functions to retrieve notation positions
- Helper functions to retrieve note time relative to beats per a minute
- Support for multiple OSMD rendering in a page
- Performance optimizations to load faster
- A react component that can be used to embed it, rather than requiring me to build my own
    - Or at least guidance on how to do this with best practices in mind
    - I currently have an issue where it blows up because I remove the component that renders it to display something, and then re-render OSMD afterwards, and I'm not quite sure how to solve this

I'm sure there's more, but this is just off the top of my head.


# Final Thoughts

I just want to give a shout out to @sschmidTU and the whole team working on this.

My goal has been to create a tool that will help me learn Piano easily and manages my learning so I don't have to.

OSMD has removed all the hard parts of figuring out how to render sheet music so I can build interactive tooling around it to support learners IN THE ACTUAL CONTEXT OF SHEET MUSIC. This is huge! And I really appreciate what it has allowed me to do, which is focusing on solving an education problem and not a technical problem.

Thank you everyone!

And come check out Piano Gym!

**Reply by sschmidTU** [reply]
Regarding note time(ing), have you seen our Wiki page for it?
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/wiki/Tutorial:-Extracting-note-timing-for-playing

Support for multiple OSMD instances on one page is already available in OSMD since a few versions, each OSMD instance has its own EngravingRules object now, no more static shared objects. See #559

Regarding React, have you seen our React component? It's just a proof of concept example project, but it works:
https://github.com/opensheetmusicdisplay/react-opensheetmusicdisplay

Performance optimization is definitely high on the list, see #937.

Wishing you all the best for Piano Gym and your endeavors, it's looking very inspired, as i said on [our Discord server](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/927#issuecomment-766828339) :)

**Reply by loganknecht** [reply]
RE: Timing - I have! That's what got me started, and it's been wonderful. I think it would be nice to have something baked in for that as well though. I've had to build a whole set of helper libraries around OSMD, which is totally fine. It's just that you all actually write tests and I don't 😂 

RE: React Component - I did look at that, and I'm not gonna lie. There's a whole slew of things that are keeping me from solving this, and I'm not sure when I'll get to circle back on it, but I am glad it's there!

**Reply by loganknecht** [reply]
If anyone comes by and wants to see a more polished video explaining Piano Gym feel free to check this video out! ❤️ 

https://www.youtube.com/watch?v=faxNDhOjlh4

**Comment by rvilarl** [comment]
Hi everyone! My name is Rodrigo Vilar. Why I am here? I always wanted to play piano. Now with Corona, I decided to buy a Yamaha P-125 which came with a flowkey subscription (www.flowkey.com) I did like the concept MIDI Piano / App helping to learn songs. I also did like the amount of MusicXML scores in the net. I suggested the flowteam to allow loading these scores and they did not like the idea so I built it myself (www.michaelecke.com/pianoplay) This idea took me to OpenSheetMusicDisplay and VexFlow. I like both projects a lot and I am happy to collaborate where I can.
